google-cloud-pubsub/lib/google/cloud/pubsub/subscriber/async_stream_pusher.rb

```diff
@@ -119,25 +127,34 @@ def stopped?
             synchronize { @stopped }
           end
 
+          def batch_nil?
+            synchronize { @batch.nil? }
+          end
+
           protected
 
           def run_background
-            synchronize do
-              until @stopped
-                if @batch.nil?
-                  @cond.wait
+            until stopped?
+              if batch_nil?
+                synchronize do
```

Comment:
Breaking up the synchronize block makes this a lot harder to reason about, because the invariants are less clear (and probably a lot less strong). I strongly recommend we keep synchronize around the entire method, as before.

Reply:
This file isn't being used anymore, but I agree. We don't have any reason to break it up like we do in Stream. Updated.